test/Mono.Linker.Tests.Cases/RequiresCapability/RequiresInCompilerGeneratedCode.cs

```diff
@@ -1121,12 +1254,98 @@ static void TestGenericTypeParameterRequirement<TUnknown> ()
 				};
 			}
 
+			[ExpectedWarning ("IL2026")]
+			[ExpectedWarning ("IL3002", ProducedBy = ProducedBy.Analyzer)]
+			[ExpectedWarning ("IL3050", ProducedBy = ProducedBy.Analyzer)]
+			static void TestSuppressionOnLambda ()
+			{
+				var lambda =
+				[RequiresUnreferencedCode ("Suppress in body")]
+				[RequiresAssemblyFiles ("Suppress in body")]
+				[RequiresDynamicCode ("Suppress in body")]
+				() => MethodWithRequires ();
+
+				lambda (); // This will produce a warning since the lambda has Requires on it
+			}
+
+			[ExpectedWarning ("IL2026")]
+			[ExpectedWarning ("IL3002", ProducedBy = ProducedBy.Analyzer)]
+			[ExpectedWarning ("IL3050", ProducedBy = ProducedBy.Analyzer)]
+			static void TestSuppressionOnLambdaWithNestedLambda ()
+			{
+				var lambda =
+				[RequiresUnreferencedCode ("Suppress in body")]
+				[RequiresAssemblyFiles ("Suppress in body")]
+				[RequiresDynamicCode ("Suppress in body")]
+				() => {
+					// The linker doesn't have enough information to associate the RUC on lambda
+					// with this nested lambda.
+					var nestedLambda =
+					[ExpectedWarning ("IL2026", ProducedBy = ProducedBy.Trimmer)]
+					() => MethodWithRequires ();
```

Comment:
The issue is that with local functions the IL reference (if there is one at all) doesn't tell us where it is defined. So `M1` and `M2` produce effectively the same [IL](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQMwAJboMLoN7LpGYAs6AsnABQCU+hxjUZA8gK4AuApgE631JGQ4gEkAdmN60A3A2GM58os3TjJfOgUFLiAX0WN92okYXHSFGPy1KV7bhoE7REqTVnnDB4t+Vk1bk7ypnrIukA=):
```csharp
public class C {
    void M1() {
        void Outer() {
            Inner();
            
            void Inner() {
            }
        }
    }
    
    void M2() {
        void Outer() {
            Inner();
        }
        
        void Inner() {
        }
    }
}
```

It might be possible to do better for cases where they capture enclosing state - there we could potentially infer the nesting from what got captured - but this would mean more reliance on compiler implementation details.

This is only really an issue for local functions though. You're right that we could do better for lambdas since the IL will always have an actual reference to the generated lambda method at the point where it is defined (I'll update the comment to reflect this). It's debatable whether we should implement better understanding of nested lambdas, or leave lambdas with the same limitation for consistency with local functions, what do you think?

Reply:
I didn't realize that with local functions you can have them "cross scope" like this. You're right, this is probably not easily solvable. I think in that case, we just need to make sure that we don't omit warnings which should be there. I don't think we will with the current implementation (in this PR), but maybe give that a thought. It's better if we produce warnings which should not be there, than the other way round.